finality-aleph/src/sync/select_chain.rs

```diff
@@ -0,0 +1,112 @@
+use std::sync::Arc;
+
+use log::{error, info, warn};
+use network_clique::SpawnHandleT;
+use parking_lot::RwLock;
+use sp_consensus::{Error, SelectChain};
+use sp_runtime::traits::Block as BlockT;
+use tokio::sync::mpsc::{channel, error::TrySendError, Receiver, Sender};
+
+use crate::{Block, BlockHash, SpawnHandle};
+
+// It never should grow that much
+const CHANNEL_SIZE: u32 = 1024;
+const LOG_TARGET: &str = "aleph-select-chain";
+
+pub fn select_chain_state_handler<B: Block>(
+    select_chain: FavouriteSelectChain<B>,
+    spawn_handle: &SpawnHandle,
+) -> SelectChainStateHandler<B::Header> {
+    let (tx, handler) = SelectChainStateHandler::new();
+
+    let favourite = select_chain.favourite;
+
+    let mut state_writer = StateWriter::<B> {
+        events: tx,
+        favourite,
+    };
+
+    spawn_handle.spawn("aleph/state_writer", async move {
+        info!(target: LOG_TARGET, "starting up StateWriter");
+        state_writer.run().await;
+        error!(target: LOG_TARGET, "StateWriter unexpectly finished");
+    });
+
+    handler
+}
+
+pub struct SelectChainStateHandler<Header> {
+    favourite_sender: Sender<Header>,
+}
+
+impl<Header> SelectChainStateHandler<Header> {
+    pub fn new() -> (Receiver<Header>, Self) {
+        let (rx, tx) = channel(CHANNEL_SIZE as usize);
+
+        (
+            tx,
+            Self {
+                favourite_sender: rx,
+            },
+        )
+    }
+    pub fn update_favourite(&self, new_favourite: Header) {
+        if let Err(TrySendError::Full(_)) = self.favourite_sender.try_send(new_favourite) {
+            warn!(target: LOG_TARGET, "SelectChainStateHandler channel full, skipping one notification")
+        }
+    }
+}
+
+pub struct StateWriter<B: Block> {
+    events: Receiver<<B as Block>::Header>,
+    favourite: Arc<RwLock<<B as Block>::Header>>,
```

Comment:
Welp, all of this is incredibly convoluted, channel to separate task which updates a rwlocked value. This also made you add a `_tx` to all tests and ignore it, which kinda sucks.

Alternative proposal – use a (probably async?) `RwLock` directly, hold and update it in `Service` and have `Forest` and `Handler` return the updated favourite header whenever needed. This should cut down on components needed and make testing of this feature in `Forest` and `Header` more natural.

Reply:
`Handler` in general is not supposed to be doing any IO operations, whether sync or async, to make testing simpler. In particular, if you were to test the `_tx` you would have to read from it rather than just examine the return values.

Considering that the `SelectChain` trait has async methods it's users might expect to be able to give away control whenever blocking on these, so using an async lock would be advisable anyway.

Reply:
After offline discussion changed the design to use oneshot channel. I think it looks quite nice now :)